coc/events.py

```diff
@@ -363,6 +365,115 @@ async def wrapped(cached_war: ClanWar, war: ClanWar, callback):
+
+        async def wrapped(cached_raid: RaidLogEntry, raid: RaidLogEntry, callback):
+            if cached_raid.attack_log != raid.attack_log:
+                changed_districts = []
+                for count, clan in enumerate(raid.attack_log):
+                    if count > len(cached_raid.attack_log):
+                        for district in clan.districts:
+                            changed_districts.append(district)
+                    else:
+                        cached_clan = cached_raid.attack_log[count]
+                        for cached_district, district in zip(cached_clan.districts, clan.districts):
+                            if cached_district.destruction != district.destruction:
+                                changed_districts.append(district)
+
+                for clan, district in changed_districts:
+                    await callback(district, raid)
+
+        return _ValidateEvent.shortcut_register(wrapped, tags, custom_class, retry_interval, RaidEvents.event_type)
+
+    @classmethod
+    def defense_district_destruction_change(cls, tags=None, custom_class=None, retry_interval=None):
+        """Event for when a district is destroyed on defense."""
+        async def wrapped(cached_raid: RaidLogEntry, raid: RaidLogEntry, callback):
+            if cached_raid.defense_log != raid.defense_log:
+                changed_districts = []
+                for count, clan in enumerate(raid.defense_log):
+                    clan.index
+                    if count > len(cached_raid.defense_log):
+                        for district in clan.districts:
+                            changed_districts.append(district)
+                    else:
+                        cached_clan = cached_raid.defense_log[count]
+                        for cached_district, district in zip(cached_clan.districts, clan.districts):
+                            if cached_district.destruction != district.destruction:
+                                changed_districts.append(district)
+
+                for clan, district in changed_districts:
+                    await callback(district, raid)
+
+        return _ValidateEvent.shortcut_register(wrapped, tags, custom_class, retry_interval, RaidEvents.event_type)
+
+    @classmethod
+    def member_attack(cls, tags=None, custom_class=None, retry_interval=None):
+        """Event for when a member attacks."""
+
+        async def wrapped(cached_raid: RaidLogEntry, raid: RaidLogEntry, callback):
+            if cached_raid.members:
+                member_attacks = (member for member in raid.members if member not in set(cached_raid.members) or member.attacks != cached_raid.get_member(member.tag).attacks)
+            else:
+                member_attacks = raid.members
+
+            for member in member_attacks:
+                await callback(member, raid)
+
+        return _ValidateEvent.shortcut_register(wrapped, tags, custom_class, retry_interval, RaidEvents.event_type)
+
+
```

Comment:
I will leave this here as a reminder that I did not thoroughly check all of this and need to do it later.